pkg/reconciler/cache/replication/replication_reconcile.go

```diff
@@ -19,102 +19,126 @@ package replication
-	var cacheApiExport *apisv1alpha1.APIExport
-	var localApiExport *apisv1alpha1.APIExport
-	var cluster logicalcluster.Name
-	var namespace string
-	var apiExportName string
-	var err error
-	cluster, namespace, apiExportName, err = kcpcache.SplitMetaClusterNamespaceKey(key)
+func (c *controller) reconcileObject(ctx context.Context,
+	key string, gvr schema.GroupVersionResource, gvk schema.GroupVersionKind,
+	retriveCacheObject func(gvr schema.GroupVersionResource, cluster, namespace, name string) (interface{}, error),
+	retriveLocalObject func(key string) (interface{}, error)) error {
+	cluster, namespace, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
 	if err != nil {
 		return err
 	}
-	cacheApiExport, err = c.retrieveApiExport(&gvr, cluster.String(), namespace, apiExportName)
+	cacheObject, err := retriveCacheObject(gvr, cluster.String(), namespace, name)
 	if err != nil && !errors.IsNotFound(err) {
 		return err
 	}
-	localApiExport, err = c.localApiExportLister.Get(key)
+	localObject, err := retriveLocalObject(key)
 	if err != nil && !errors.IsNotFound(err) {
 		return err
 	}
 	if errors.IsNotFound(err) {
-		// issue a live GET to make sure the localApiExport was removed
-		unstructuredLiveLocalApiExport, err := c.dynamicLocalClient.Cluster(cluster).Resource(gvr).Get(ctx, apiExportName, metav1.GetOptions{})
+		// issue a live GET to make sure the localObject was removed
+		_, err = c.dynamicLocalClient.Cluster(cluster).Resource(gvr).Namespace(namespace).Get(ctx, name, metav1.GetOptions{})
 		if err == nil {
-			return fmt.Errorf("the informer used by this controller is stale, the following APIExport was found on the local server: %s/%s/%s but was missing in the informer", cluster, unstructuredLiveLocalApiExport.GetNamespace(), unstructuredLiveLocalApiExport.GetName())
+			return fmt.Errorf("the informer used by this controller is stale, the following %s resource was found on the local server: %s/%s/%s but was missing from the informer", gvr, cluster, namespace, name)
 		}
 		if !errors.IsNotFound(err) {
 			return err
 		}
 	}
 
-	var unstructuredCacheApiExport *unstructured.Unstructured
-	var unstructuredLocalApiExport *unstructured.Unstructured
-	if cacheApiExport != nil {
-		unstructuredCacheApiExport, err = toUnstructured(cacheApiExport)
+	var unstructuredCacheObject *unstructured.Unstructured
+	var unstructuredLocalObject *unstructured.Unstructured
+	if cacheObject != nil && !reflect.ValueOf(cacheObject).IsNil() {
+		unstructuredCacheObject, err = toUnstructured(cacheObject)
 		if err != nil {
 			return err
 		}
-		unstructuredCacheApiExport.SetKind("APIExport")
-		unstructuredCacheApiExport.SetAPIVersion(gvr.GroupVersion().String())
+		unstructuredCacheObject.SetKind(gvk.Kind)
+		unstructuredCacheObject.SetAPIVersion(gvr.GroupVersion().String())
 	}
-	if localApiExport != nil {
-		unstructuredLocalApiExport, err = toUnstructured(localApiExport)
+	if localObject != nil && !reflect.ValueOf(localObject).IsNil() {
```

Comment:
an interface is a two-word data, it holds data and a type, since this code operates on the `interface` type and informers are strongly typed we need to make sure that the data part is not nil.